### DISCUSSION TITLE
Is it possible to automatically port over a GUI application from Linux/Windows to Android that was written in Rust?

### DISCUSSION THREAD

**Discussion by Raj2032** [rust-windowing/winit#2111] [General]
Hey guys,

So I am wondering I am thinking of using [`conrod`](https://github.com/PistonDevelopers/conrod) GUI library, however I _don't_ think it supports Android.

Is it possible to some how use `winit` to make it work on Android?

If this is the case, say my GUI application worked successfully on Linux, would it work on `winit` or am I expecting to have breakages?

Also I wanted to know do I have to keep fiddling around with `winit` every time I make a change with my `conrod` application?

**Comment by madsmtm** [comment]
[`conrod_winit`](https://github.com/PistonDevelopers/conrod/tree/master/backends/conrod_winit) exist, so that could _maybe_ be used to make it work on Android, though you'll have to ask the `conrod` developers about that - depends on how whether they use the system rendering primitives, or roll their own.

**Comment by rib** [comment]
Currently one of the difficulties with writing winit applications that need to run on Android is that it's the only platform that requires you to hook into `Resumed` and `Suspended` events to know when you have a valid native window which can be used for rendering, and where you have to re-create a surface for rendering each time the application resumes.

The backend will actually panic if you try to access the native_window while the application is paused which may seem brutal but the OS actually destroys the underlying native window and creates a new one when the app resumes which needs to be handled by applications at a higher level.

This shouldn't necessarily be a big problem to handle but as it is currently then most winit integrations are following some common patterns that tend to just assume that a window can be created at the same time as an event loop and all graphics state can also be immediately initialized, and this will basically lead to a panic on Android.

I've recently been experimenting with an alternative Android glue layer (https://github.com/rib/agdk-rust/tree/main/game-activity) and updated winit backend (https://github.com/rib/winit/tree/agdk-game-activity) where I've then also been making a few test applications that demonstrate running on both Android and desktop here: https://github.com/rib/agdk-rust/tree/main/examples.

This stuff is still work-in-progress though.

Hopefully though it's possible to iterate on some of the integration layers and on winit itself to make it easier to write applications that Just Work across mobile/Android and desktop. E.g. I sent this initial PR for egui today to try and improve Android support there: https://github.com/emilk/egui/pull/1634 and maybe similar changes can be made for `conrod_winit` too.
